File: 20.valid-parentheses.java
/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        if(s.length() % 2 != 0) return false;

        Deque<Character> q = new ArrayDeque<>(s.length());
        char c;

        for(int i = 0; i < s.length(); i++) {
            switch(s.charAt(i)) {
                case '(':
                case '[':
                case '{':
                    q.push(s.charAt(i));
                    break;

                case ')':
                    if(q.size() == 0 || q.pop() != '(') return false;
                    break;
                case ']':
                    if(q.size() == 0 || q.pop() != '[') return false;
                    break;
                case '}':
                    if(q.size() == 0 || q.pop() != '{') return false;
                    break;
            } 
        }

        return q.size() == 0;
    }
}
// @lc code=end

File: cpp/20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
public:

    char toclosed(char c) {
        switch(c) {
            case '(':
                return ')';
            case '[':
                return ']';
            case '{':
                return '}';
            default:
                return '\0';
        }
    }

    bool isValid(string s) {
        std::stack<char> stk;

        for(char c : s) {
            if(c == '(' | c == '[' | c == '{') {
                stk.push(toclosed(c));
            }
            else if(stk.empty() || c != stk.top()) {
                return false;
            }
            else {
                stk.pop();
            }
        }

        return stk.empty();
    }
};
// @lc code=end

